File: apps/backoffice/lib/api/onboarding.ts
import 'server-only';

import { cache } from 'react';

import { apiRequest } from '@/lib/api/client';

// --- Raw types (API JSON shape, snake_case) ---

interface OnboardingRaw {
  flow_version: number;
  status: OnboardingStatus;
  current_step: OnboardingStepKey;
  steps: Partial<Record<OnboardingStepKey, OnboardingStepStatus>>;
  completed_at: string | null;
}

// --- Frontend types (camelCase) ---

export type OnboardingStatus = 'IN_PROGRESS' | 'COMPLETED' | 'DISMISSED';
export type OnboardingStepStatus = 'COMPLETED' | 'SKIPPED';
export type OnboardingStepKey =
  | 'WELCOME'
  | 'BRAND'
  | 'FIRST_BRANCH'
  | 'CATALOG_UPLOAD'
  | 'CATALOG_REVIEW'
  | 'TEAM'
  | 'COMPLETE';

export interface Onboarding {
  flowVersion: number;
  status: OnboardingStatus;
  currentStep: OnboardingStepKey;
  steps: Partial<Record<OnboardingStepKey, OnboardingStepStatus>>;
  completedAt: string | null;
}

// --- Mappers ---

function mapOnboarding(raw: OnboardingRaw): Onboarding {
  return {
    flowVersion: raw.flow_version,
    status: raw.status,
    currentStep: raw.current_step,
    steps: raw.steps,
    completedAt: raw.completed_at,
  };
}

// --- API functions ---

export const getOnboarding = cache(async (): Promise<Onboarding> => {
  return mapOnboarding(
    await apiRequest<OnboardingRaw>({ path: '/v1/onboarding', branchScoped: false }),
  );
});

File: apps/api/internal/domain/onboarding.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// OnboardingStatus is the account setup lifecycle.
type OnboardingStatus string

const (
	OnboardingStatusInProgress OnboardingStatus = "IN_PROGRESS"
	OnboardingStatusCompleted  OnboardingStatus = "COMPLETED"
	OnboardingStatusDismissed  OnboardingStatus = "DISMISSED"
)

// OnboardingStepStatus records how an account resolved one setup step.
type OnboardingStepStatus string

const (
	OnboardingStepStatusCompleted OnboardingStepStatus = "COMPLETED"
	OnboardingStepStatusSkipped   OnboardingStepStatus = "SKIPPED"
)

// OnboardingStepKey is a stable identifier whose order is decided by the current flow version.
type OnboardingStepKey string

const (
	OnboardingStepWelcome       OnboardingStepKey = "WELCOME"
	OnboardingStepBrand         OnboardingStepKey = "BRAND"
	OnboardingStepFirstBranch   OnboardingStepKey = "FIRST_BRANCH"
	OnboardingStepCatalogUpload OnboardingStepKey = "CATALOG_UPLOAD"
	OnboardingStepCatalogReview OnboardingStepKey = "CATALOG_REVIEW"
	OnboardingStepTeam          OnboardingStepKey = "TEAM"
	OnboardingStepComplete      OnboardingStepKey = "COMPLETE"
)

// Onboarding is the versioned setup state for one account.
type Onboarding struct {
	ID          uuid.UUID
	AccountID   uuid.UUID
	FlowVersion int
	Status      OnboardingStatus
	CurrentStep OnboardingStepKey
	Steps       map[OnboardingStepKey]OnboardingStepStatus
	CompletedAt *time.Time
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// OnboardingProgress resolves one step and points at the next screen to resume.
type OnboardingProgress struct {
	Step        OnboardingStepKey
	StepStatus  OnboardingStepStatus
	CurrentStep OnboardingStepKey
}
